src/content/blog/ddev-diffy-introduction.md

```diff
@@ -0,0 +1,66 @@
+---
+title: "Diffy: Anatomy of an Advanced DDEV Add-on"
+pubDate: 2024-08-03
+modifiedDate: 2024-08-03
+summary: 'Some advanced topics: saving configuration variables, "npm install" inside of the container, building container for multiple architectures'
+author: Yuri Gerasymov
+featureImage:
+  src: /img/blog/2024/08/ddev+diffy.png
+  alt: DDEV and Diffy integration
+  credit: ''
+categories:
+  - TechNotes
+---
+
+[Diffy](https://diffy.website) is a visual regression testing tool that allows you to take screenshots of your website and compare them so you know exactly what pages got changed and where. We decided to build DDEV integration so it is possible to take screenshots from the local website, upload them to Diffy, and then compare them with screenshots from other environments.
+
+Watch [3 mins video](https://www.loom.com/share/1944eb3462934279a175e65e16e715ed) or read [documentation](https://docs.diffy.website/features/local-development/ddev-add-on).
+
+Architecturally integration consists of:
+
+* Docker container with preinstalled node and Chrome 111
+* Node.js app that runs Chrome
+* The app itself is hosted in a separate repository as it is used in other places (like production for the main service itself).
+
+Let’s go through the way integration is built.
+
+## Download node app
+To download the app we use `install.yaml` post_install_actions:
+```
+- rm -rf diffy-worker && mkdir diffy-worker
+- docker run -it --rm -v ./diffy-worker:/diffy-worker --user $DDEV_UID:$DDEV_GID ddev/ddev-utilities bash -c "cd /diffy-worker && wget -qO- https://github.com/DiffyWebsite/diffy-worker/archive/refs/heads/main.tar.gz | tar xz --strip-components=1"
+```
+
+This is a particularly interesting approach as we use `ddev-utilities` to download the code and unpack it.
+
+## App configuration
+App requires an API key from the service and also project ID. This is a pretty interesting setup that we borrowed from the Platformsh project.
+We keep the API key in the global configuration (i.e. `~/.ddev/global_config.yaml`) and project ID is in local configuration (your project’s `.ddev/config.yaml`).
+
+First we check if variables are set and if they are not — we request the values (`read token`) and then save them with `ddev config global --web-environment-add` in case of global variable and `ddev config --web-environment-add` in case of project based variable.
+
+Pay attention that ‘{{ }}’ are [go templates](https://ddev.readthedocs.io/en/stable/users/extend/additional-services/#template-action-replacements-advanced) and not bash code.
```

Comment:
```suggestion
Note that ‘{{ }}’ are [go templates](https://ddev.readthedocs.io/en/stable/users/extend/additional-services/#template-action-replacements-advanced) and not bash code.
```

Reply:
No longer relevant. I removed this section completely.